handshake/src/providers/discogs.ts

```diff
@@ -0,0 +1,281 @@
+}> {
+  const url = new URL(`https://api.discogs.com/oauth/request_token`);
+
+  // https://www.discogs.com/developers#header-2.-send-a-get-request-to-the-discogs-request-token-url
+  url.searchParams.set("oauth_nonce", nonce);
+  url.searchParams.set("oauth_consumer_key", consumerKey);
+  // ATTENTION notice the required ampersand at the end.
+  url.searchParams.set("oauth_signature", consumerSecret + "&");
+  url.searchParams.set("oauth_signature_method", "PLAINTEXT");
+  url.searchParams.set("oauth_timestamp", "" + Date.now());
+  url.searchParams.set("oauth_callback", callbackHandlerUrl);
+
+  let res;
+  try {
+    res = await fetch(url, {
+      method: "GET",
+      headers: {
+        "User-Agent": "handshake/1.0",
+      },
+    });
+  } catch (e) {
+    throw new UnknownProviderError(`FETCH url=${url}: ${e}`);
+  }
+
+  if (!res.ok) {
+    const text = await res.text();
+    error(`Discogs returned status=${res.status} url=${url} text=${text}`);
+    // TODO expand expected errors.
+
+    if (text.includes("Invalid consumer.")) {
+      throw new OAuthCallbackError("invalid_client", `Wrong consumer secret.`);
+    }
+
+    // Never return the error to the user. Discogs may include the consumer
+    // secret in the response. 😱😱
+    if (text.includes("Expected signature base string.")) {
+      throw new OAuthCallbackError("invalid_client", `Wrong consumer secret.`);
+    }
+
+    throw new UnknownProviderError(
+      `Unexpected error from Discogs: status=${res.status} url=${url} text=${text}`,
+    );
+  }
+
+  const text = await res.text();
+  debug("Received from Discogs", text);
+  const json = Object.fromEntries(new URLSearchParams(text));
+
+  let parsed;
+  try {
+    parsed = z
+      .object({
+        oauth_token: z.string(),
+        oauth_token_secret: z.string(),
+        oauth_callback_confirmed: z.enum(["true", "false"]),
+      })
+      .parse(json);
+  } catch (e) {
+    error("Failed to parse response from Discogs", text);
+    throw new UnknownProviderError(
+      `Failed to parse response from Discogs e: ${e}`,
+    );
+  }
+
+  return {
+    requestToken: parsed.oauth_token,
+    requestTokenSecret: parsed.oauth_token_secret,
+    callbackConfirmed: parsed.oauth_callback_confirmed === "true",
+  };
+}
+
+async function requestAccessToken(
+  consumerKey: string,
+  consumerSecret: string,
+  tokenSecret: string,
+  oauthToken: string,
+  verifier: string,
+  nonce: string,
+) {
+  const url = `https://api.discogs.com/oauth/access_token`;
+
+  // https://www.discogs.com/developers#page:authentication,header:authentication-oauth-flow
+  let res;
+  try {
+    res = await fetch(url, {
+      method: "POST",
+      headers: {
+        "User-Agent": "handshake/1.0",
+        Authorization: "OAuth",
+      },
+      body: new URLSearchParams({
+        oauth_nonce: nonce,
+        oauth_token: oauthToken,
+        oauth_consumer_key: consumerKey,
+        oauth_signature: consumerSecret + "&" + tokenSecret,
```

Comment:
good find! I think it is not documented (by Discogs) that this field uses these two values